core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -44,7 +44,14 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
     blockManager.get(key) match {
       case Some(blockResult) =>
         // Partition is already materialized, so just return its values
+        val existingMetrics = context.taskMetrics.inputMetrics
+        val prevBytesRead = existingMetrics
+          .filter(_.readMethod == blockResult.inputMetrics.readMethod)
+          .map(_.bytesRead)
+          .getOrElse(0L)
```

Comment:
What if there are 3 input sources that interleave here?  Suppose you have (1) input from cache, (2) input from Hadoop, and (3) input from cache.  My understanding is that when (2) starts being read, it will clobber the input metrics from (1).  Then, when (3) is read, it will again clobber the input metrics, so the metrics won't properly reflect the total data read from cache (they'll only reflect the data read from (3)).  Is that right?